content/silver/concepts/automatic-attributes.md

```diff
@@ -159,11 +162,229 @@ top::Stmt ::= lifted::Decls
+
+```
+destruct attribute compareTo;
+
+nonterminal Type;
+
+attribute compareTo occurs on Type;
+
+abstract production intType
+top::Type ::=
+{
+  propagate compareTo;
+}
+
+abstract production dataType
+top::Type ::= name::String
+{
+  propagate compareTo;
+}
+
+abstract production fnType
+top::Type ::= a::Type b::Type
+{
+  propagate compareTo;
+}
+
+```
+
+A destruct attribute might be used for comparing types in a functional language implementation.  The above translates to the following equivalent specification:
+
+```
+inherited attribute compareTo<a (i::InhSet)>::Decorated a with i;
+
+nonterminal Type;
+
+attribute compareTo<Type {}> occurs on Type;
+
+abstract production intType
+top::Type ::=
+{
+  -- No children for which to propagate the attribute
+}
+
+abstract production dataType
+top::Type ::= name::String
+{
+  -- No children for which to propagate the attribute
+}
+
+abstract production fnType
+top::Type ::= a::Type b::Type
+{
+  a.compareTo =
+    case top.compareTo of
+    | fnType(a2, b2) -> a2
+    | _ -> error("Destruct attribute compareTo demanded on child a of production fnType when given value doesn't match")
+    end;
+  b.compareTo =
+    case top.compareTo of
+    | fnType(a2, b2) -> b2
+    | _ -> error("Destruct attribute compareTo demanded on child a of production fnType when given value doesn't match")
+    end;
+}
+```
+
+The type of a destruct attribute is a [`Decorated` (reference) type](/silver/concepts/decorated-vs-undecorated/), such that attributes may be accessed on the corresponding tree.  The type of the destructed tree and the inherited attributes carried by the reference are specified as two type parameters on the attribute, of kind * and [InhSet](/silver/concepts/decorated-vs-undecorated/#inhset-types).
+
+Destruct attributes provide an overload for attribute occurrence such that `occurs on` for a destruct attribute with no type arguments will forward to an attribute occurrence with the nonterminal and `{}` (the empty set of inherited attributes) provided as the type arguments.  Often one wishes to make use of inherited attributes on the deconstructed tree; if a single type argument is provided, it will be used as the set of inherited attributes for the reference.  For example writing
+```
+attribute compareTo<{someInh}> occurs on Type;
+```
+will forward to
+```
+attribute compareTo<{someInh} Type> occurs on Type;
+```
+giving a type `Decorated Type with {someInh}`; thus we could access `a.compareTo.someInh` in the `fnType` production.
+
+Destruct attributes are typically used in conjunction with equality or ordering attributes, described next.
+
+# Equality attributes
```

Comment:
Wouldn't it flow better to pose the problem of checking for equality first and talk about the strategy of passing in as an inherited attribute the tree being compared to the one that is decorated? I think big picture is missing here. The discussion of `destruct` attributes seems not well motivated.

Reply:
I don't see where you discuss the problem of checking equality of trees and *then* say that destruct and equality attributes are used to solve this problem. I think it is helpful to say what the general idea is - of passing down the tree that is a tree is being compared to - and then deconstructing it as one works down the structure of the tree.

I know this is supposed to be a "reference manual" - but it seems that nice description of the problem would fit well for each of these propagated attributes.